File: Python3/6. Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        rows = [''] * numRows
        backwards = True
        row = 0

        for c in s:
            rows[row] += c
            if row == 0 or row == numRows - 1:
                backwards = not backwards
            
            if backwards:
                row -= 1
            else:
                row += 1
            
        return ''.join(rows)

File: Java/6. Zigzag Conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        String[] rows = new String[numRows];
        Arrays.fill(rows, "");
        int row = 0;
        boolean backwards = true;

        for (char c: s.toCharArray()) {
            rows[row] += c;
            
            if (row == 0 || row == numRows - 1) {
                backwards = !backwards;
            }

            if (backwards) {
                row--;
            } else {
                row++;
            }
        }
        

        String ans = new String();
        for (String sb: rows) {
            ans += sb.toString();
        }

        return ans;
    }
}